### PR TITLE
final version

### PR DESCRIPTION
**Pull request by cyb3rel**

### CODE REVIEW

File: CMP1903_A1_2324/Die.cs [hunk 1]
```diff
@@ -13,11 +13,12 @@ internal class Die
...
-
+        private static Random random = new Random();
+        public int diceValue { get; set; } //stores values after rolling dice
+        public int Roll() //roll dice
+        {
+            diceValue = random.Next(1, 7); //generates a random number between 1 and 6
+            return diceValue;
+        }
     }
 }
```
Comment (EJohnston2) [inline CMP1903_A1_2324/Die.cs:24]: Die class has met the assignment requirements, it simulates a die that is assigned a random value between 1 and 6. Easy to follow and read, meets requirements of C#. Only thing that could be improved is using comments to make it easier to follow.

File: CMP1903_A1_2324/Game.cs [hunk 2]
```diff
@@ -15,7 +15,27 @@ internal class Game
...
+            Console.WriteLine("Die 2 = " + die2.diceValue);
+            Console.WriteLine("Die 3 = " + die3.diceValue); //outputs the individual dice rolls
+
+            diceTotal += die1.diceValue + die2.diceValue + die3.diceValue;
+            Console.WriteLine("Dice total = " +  diceTotal);
+
+            return diceTotal; //returns total
+        }
     }
 }
```
Comment (EJohnston2) [inline CMP1903_A1_2324/Game.cs:41]: Meets requirements, runs as it should and meets the standards of C#, good use of comments throughout the code explaining the different processes being performed.

File: CMP1903_A1_2324/Testing.cs [hunk 3]
```diff
@@ -14,7 +16,17 @@ internal class Testing
...
+            Debug.Assert(test == (game.die1.diceValue + game.die2.diceValue + game.die3.diceValue),
+                 "dice total is incorrect");
 
-        //Method
+            Die dice = new Die();
+            int testDie = dice.Roll(); //creates a die object to be tested
+            Debug.Assert(testDie >= 1 && testDie <= 6, "die is incorrect"); //checks whether an individual dice roll is within range
+        }
     }
 }
```
Comment (EJohnston2) [inline CMP1903_A1_2324/Testing.cs:32]: meets the requirement for the testing class, uses appropriate outputs for an invalid value and accurately tests the boundaries. good show of error handling.

Comment (EJohnston2) [reply CMP1903_A1_2324/Testing.cs:32]: only improvements could be a few more comments, especially in the die class to make the code simpler to read through and understand.

**Review by sorioxis — COMMENTED** [review]
Hi Elonnah - here is my review of your code. I'll run it down the assessment checklist as per the brief:

### Have the requirements been met? ###

Yes, you have met all the requirements that were listed in the brief. All of the classes included perform as expected and have all the necessary methods that are listed in the assessment brief. I can see that your code has a Die class, Game class and Testing class, which were the minimum requirements in the brief. You have also utilised Roll(); correctly, which was a requirement.

I can see in your code that the game class correctly totals the dice values - you have also added testing to confirm that the sum of the die rolls are correct and between 1-6. This was listed as a debug requirement in the brief, so good job.

**Die Class:**

- [x] Has a property that stores the value of the die
- [x] Generates random number between 1-6
- [x] Has the Roll(); method

**Game Class:**

- [x] Creates 3 dice objects
- [x] Has a method that rolls the die
- [x] Outputs individual dice rolls
- [x] Outputs the total of the dice rolls

**Program Class**

- [x] Creates a new game object
- [x] Creates a new testing object
- [x] Calls the methods from other classes 
- [x] Has ReadKey(); to ensure the program doesn't close instantly

**Testing Class:**

- [x] Creates a new game object
- [x] Rolls dice
- [x] Ensures dice rolls are between 1 and 6
- [x] Ensures roll sum meets expected result

### Is the code formatted correctly? ###

All of your variables and methods are given appropriate names and follow common naming conventions. You have utilised comments where necessary to add extra readability for your code. You have also used the correct parameters for methods - I can see that a lot of your methods are public, which is great for this specific program as you will be calling them in your Program class.

### Is the code easy to read? ###

Yes, your commenting helps a ton in increasing the readability of the code. You have also effectively indented your code, which is great for reading it over / the flow of the code. In the future, you could maybe add xml comments to your code - summary, returns, etc, which would further increase readability, but as it is, it suits the program and the requirements of the brief.

### Are different errors handled correctly? ###

You have error handling included in your testing class for debugging - your code checks that the die rolls are between 1-6 and the sum meets the expected result, which is good & meets the brief requirements. Furthermore, you have correctly identified that your code closes instantly and added a ReadKey(); line to ensure that it stops closing on start. 

Overall, this is really good code. You have met the requirements of the brief completely.

**Pull request merged**